server/src/main/java/org/elasticsearch/cluster/coordination/ElectionSchedulerFactory.java

```diff
@@ -0,0 +1,203 @@
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.cluster.coordination;
+
+import org.elasticsearch.common.SuppressForbidden;
+import org.elasticsearch.common.component.AbstractComponent;
+import org.elasticsearch.common.lease.Releasable;
+import org.elasticsearch.common.settings.Setting;
+import org.elasticsearch.common.settings.Setting.Property;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.unit.TimeValue;
+import org.elasticsearch.common.util.concurrent.AbstractRunnable;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.threadpool.ThreadPool.Names;
+
+import java.util.Random;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.LongSupplier;
+
+/**
+ * It's provably impossible to guarantee that any leader election algorithm ever elects a leader, but they generally work (with probability
+ * that approaches 1 over time) as long as elections occur sufficiently infrequently, compared to the time it takes to send a message to
+ * another node and receive a response back. We do not know the round-trip latency here, but we can approximate it by attempting elections
+ * randomly at reasonably high frequency and backing off (linearly) until one of them succeeds. We also place an upper bound on the backoff
+ * so that if elections are failing due to a network partition that lasts for a long time then when the partition heals there is an election
+ * attempt reasonably quickly.
+ */
+public class ElectionSchedulerFactory extends AbstractComponent {
+
+    private static final String ELECTION_INITIAL_TIMEOUT_SETTING_KEY = "cluster.election.initial_timeout";
+    private static final String ELECTION_BACK_OFF_TIME_SETTING_KEY = "cluster.election.back_off_time";
+    private static final String ELECTION_MAX_TIMEOUT_SETTING_KEY = "cluster.election.max_timeout";
+
+    /*
+     * The first election is scheduled to occur a random number of milliseconds after the scheduler is started, where the random number of
+     * milliseconds is chosen uniformly from
+     *
+     *     (0, min(ELECTION_INITIAL_TIMEOUT_SETTING, ELECTION_MAX_TIMEOUT_SETTING_KEY)]
+     *
+     * For `n > 1`, the `n`th election is scheduled to occur a random number of milliseconds after the `n - 1`th election, where the random
+     * number of milliseconds is chosen uniformly from
+     *
+     *     (0, min(ELECTION_INITIAL_TIMEOUT_SETTING + (n-1) * ELECTION_BACK_OFF_TIME_SETTING, ELECTION_MAX_TIMEOUT_SETTING_KEY)]
```

Comment:
I like this formula. I wonder if we should try to implement ElectionScheduler exactly this way. We would "just" need to keep track of the round n (which might be useful anyhow for debugging purposes), and there would be no need for currentMaxTimeoutMillis, which feels more awkward. The only arithmetic calculation we would need to do then would be just a straight-forward implementation of this formula, which would preferably be done in a single static method at the top-level here. This would also allow us to get rid of the methods `randomPositiveLongLessThan` and `backOffCurrentMaxTimeout`, both of which could be folded into this static method.

Reply:
Hm. I did this, but there's now potential for integer overflow, so perhaps we do need that upper bound on the settings?

Also I'm unconvinced at the need for a static method to do this calculation. We test it via the actual timeouts chosen, so why extract it?

Also `randomPositiveLongAtMost` is still needed, and you wanted a test for that.

Reply:
On reflection, there was already potential for integer overflow without an upper bound on the settings and this change makes it no worse.